### PR TITLE
Player Tick events

### PR DESCRIPTION
**Pull request by AlphaMode**

### CODE REVIEW

**Comment by maityyy** [comment]
This will remove a lot of mixins, good idea ;)

**Comment by Technici4n** [comment]
After discussion on discord, it would be better to add before and after server player tick events. This pair of events could go into the existing `ServerPlayerEvents`.

**Comment by maityyy** [comment]
> After discussion on discord, it would be better to add before and after server player tick events. This pair of events could go into the existing `ServerPlayerEvents`.

PlayerEntity is only handled from the server? As far as I know, no. **Server**PlayerEvents is very confusing.

File: fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/PlayerEvents.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+import net.fabricmc.fabric.api.event.Event;
+import net.fabricmc.fabric.api.event.EventFactory;
+
+public class PlayerEvents {
+	/**
+	 * An event that is called right before the player is ticked.
+	 *
+	 * @see PlayerEntity#tick()
+	 */
+	public static final Event<BeforePlayerTick> BEFORE_PLAYER_TICK = EventFactory.createArrayBacked(BeforePlayerTick.class, callbacks -> player -> {
```
Comment (Player3324) [inline fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/PlayerEvents.java]: <https://github.com/FabricMC/fabric/blob/1.17/fabric-lifecycle-events-v1/src/main/java/net/fabricmc/fabric/api/event/lifecycle/v1/ServerTickEvents.java> introduced START_x_TICK and END_x_TICK as well as a matching theme for the callbacks. Please rename the methods accordingly.

Javadoc also needs to be specific about the the callbacks being invoked on both client and server for all players.

**Comment by Technici4n** [comment]
@AlphaMode are you planning on updating this PR?

File: fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java [hunk 2]
```diff
@@ -109,6 +132,26 @@
 		boolean allowDeath(ServerPlayerEntity player, DamageSource damageSource, float damageAmount);
 	}
 
+	@FunctionalInterface
+	public interface BeforePlayerTick {
```
Comment (Technici4n) [inline fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java]:
```suggestion
	public interface StartTick {
```

File: fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java [hunk 3]
```diff
@@ -109,6 +132,26 @@
...
+		/**
+		 * Called before the player tick.
+		 *
+		 * @param player
+		 */
+		void prePlayerTick(PlayerEntity player);
```
Comment (Technici4n) [inline fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java]:
```suggestion
		void onStartTick(ServerPlayerEntity player);
```

Comment (Technici4n) [reply fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java]: Notice the `ServerPlayerEntity`.

File: fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java [hunk 4]
```diff
@@ -72,6 +73,28 @@
 		return true;
 	});
 
+	/**
+	 * An event that is called right before the player is ticked.
+	 *
+	 * @see PlayerEntity#tick()
+	 */
+	public static final Event<BeforePlayerTick> START_PLAYER_TICK = EventFactory.createArrayBacked(BeforePlayerTick.class, callbacks -> player -> {
```
Comment (Technici4n) [inline fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java]: Should be named `START_TICK`.

File: fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java [hunk 5]
```diff
@@ -72,6 +73,28 @@
...
+			callback.prePlayerTick(player);
+		}
+	});
+
+	/**
+	 * An event that is called at the end of the player tick.
+	 *
+	 * @see PlayerEntity#tick()
+	 */
+	public static final Event<AfterPlayerTick> END_PLAYER_TICK = EventFactory.createArrayBacked(AfterPlayerTick.class, callbacks -> player -> {
```
Comment (Technici4n) [inline fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java]: `END_TICK`

File: fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java [hunk 6]
```diff
@@ -109,6 +132,26 @@
...
+		/**
+		 * Called before the player tick.
+		 *
+		 * @param player
+		 */
+		void prePlayerTick(PlayerEntity player);
+	}
+
+	@FunctionalInterface
+	public interface AfterPlayerTick {
```
Comment (Technici4n) [inline fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java]:
```suggestion
	public interface EndTick {
```

File: fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java [hunk 7]
```diff
@@ -109,6 +132,26 @@
...
+		/**
+		 * Called after the player tick.
+		 *
+		 * @param player
+		 */
+		void postPlayerTick(PlayerEntity player);
```
Comment (Technici4n) [inline fabric-entity-events-v1/src/main/java/net/fabricmc/fabric/api/entity/event/v1/ServerPlayerEvents.java]:
```suggestion
		void onEndTick(ServerPlayerEntity player);
```

**Review by Technici4n — CHANGES_REQUESTED** [review]
Naming issues, for consistency with lifecycle and other events. Otherwise LGTM.

**Comment by AlphaMode** [comment]
Closed as you can just use the Lookup api with world tick events

**Pull request closed without merging**